[FLOC-4156] Add tests for the V3 API.

sarum90 opened the pull request:
The V3 API was introduced originally without any tests. This added TestCase
tests the ICloudInstance interface, and is instantiated in a manner to test the
GCE implementation, the EC2 implementation, and the Rackspace implementation.

Credentials are acquired using acceptance.yaml rather than environment
variables.

sarum90 committed:
Add tests for the V3 API.

The V3 API was introduced originally without any tests. This added TestCase
tests the ICloudInstance interface, and is instantiated in a manner to test the
GCE implementation, the EC2 implementation, and the Rackspace implementation.

Credentials are acquired using acceptance.yaml rather than environment
variables.

Azulinho commented on README.mdown:
```diff
@@ -32,3 +32,20 @@ To run tests do:
+Note that instead of starting from that configuration, you can start with an
+empty yml file and the tests will print out what configuration you need to add
+in to your yaml file in order to run the tests.
+
```
that is really cool

sarum90 replied:
I got the idea from flocker's tests. Skipping tests with actionable messages about what you have to do to actually run the test I consider to be a really nice developer experience.

Azulinho commented on README.mdown:
```diff
@@ -32,3 +32,20 @@ To run tests do:
+In an effort to attempt to reduce the amount of work to set up the
+configuration, they build on the existing acceptance.yml as described at:
+http://doc-dev.clusterhq.com/gettinginvolved/appendix.html#acceptance-testing-configuration.
+
```
this one is really bugging me,
The code in this repo doesn't use anything from the flocker codebase. The idea is that it can be consumed horizontally to build and mange all clusterHQ microservices and fat services, without being constrained to changes happening in the flocker code base.
Considering that you already output what the yaml config should be if someone tries to run this without a valid config, then I would prefer if you remove L37-39.
pleeeeease

sarum90 replied:
Sure thing. Yea that bugged me too.

sarum90 replied:
Removed. I still call it `acceptance.yml`, and right now the configuration is compatible with a ClusterHQ flocker acceptance.yml (although it is extended to have ssh_keys). Personally I'm going to use the same file on my machine. However, with this sentence removed I no longer think we have any requirement to preserve that compatibility long-term.

Azulinho commented on bookshelf/tests/api_v3/test_cloud.py:
```diff
@@ -0,0 +1,339 @@
+    self.config: A valid configuration for the ICloudInstanceFactory provider.
+    self.distribution: The distribution that the config launches.
+    self.region: A valid region  for the ICloudInstanceFactory provider.
+    """
```
double-spacing '  for'

Azulinho replied:
I should get a bounty for this one, as I never ever catch these sort of things

sarum90 replied:
Fixed - thanks.

Azulinho commented on bookshelf/tests/api_v3/test_cloud.py:
```diff
@@ -0,0 +1,339 @@
+            security_groups=['ssh']
+        ).serialize()
+        self.distribution = Distribution.CENTOS7
+        self.region = 'us-west-2'
```
can we get the region from the config.yaml file?

we default too many of our bits and tests and things to us-west-2

sarum90 replied:
Sure... I need to think about how to get AMI then.... maybe I'll just hard code a mapping from region to AMI?

Azulinho replied:
I hate that about ubuntu, in that they change their AMIs every month or so.

I have just been deploying this file into most of my projects, but would rather be able to scrap the listing directly from the ubuntu cloud images website

I don't have a good suggestion besides maybe drop this file or something alike into the repo?
/config-examples/ec2.yaml ?

dunno really

```
\>cat ec2.yaml 
# list of cloud providers
us-east-1:
  ami: 'ami-ffe3c695'
eu-central-1:
  ami: 'ami-86564fea'
eu-west-1:
  ami: 'ami-bf72c4cc'
ap-southeast-1:
  ami: 'ami-7ac30f19'
ap-northeast-1:
  ami: 'ami-0419256a'
ap-southeast-2:
  ami: 'ami-520b2e31'
sa-east-1:
  ami: 'ami-a0b637cc'
us-west-1:
  ami: 'ami-34126654'
us-west-2:
  ami: 'ami-22b9a343'
```

Azulinho replied:
when in doubt ask google

https://pypi.python.org/pypi/UbuntuAMI

Azulinho replied:
'Author: Lx Yu' <<-- I am sending him a deep fried mars bars as a thank you

Azulinho replied:
the above doesn't work,
but the ansible ec2-ami  plugin actually shows that it might be easier than we think
https://github.com/openshift/openshift-ansible/blob/master/playbooks/aws/openshift-cluster/library/ec2_ami_find.py#L149

sarum90 replied:
cool, that's effectively what we open coded for GCE for what it's worth. They actually deprecate and remove old images though, so it seemed like a necessity. I think for now I'm just going to do a static set of AMIs. Maybe if I'm bored later I'll change the API to take an {ami_owner, ami_prefix} set of parameters and find the latest AMI.

Azulinho replied:
aye

Azulinho commented on bookshelf/tests/api_v3/test_cloud.py:
```diff
@@ -0,0 +1,339 @@
+        ).serialize()
+        self.distribution = Distribution.CENTOS7
+        self.region = 'HKG'
+        self.instance_factory = RackspaceInstance
```
HKG is a good default as not many people I know have been there, but it would be cool if I could change these from the config.yam file

sarum90 replied:
+1, I'll do this for all of them.

sarum90 replied:
Added the ability to set optional parameters in the config, and used it for a bunch of things.

Azulinho commented on bookshelf/tests/api_v3/test_cloud.py:
```diff
@@ -0,0 +1,339 @@
+        Done as one large test rather than many little ones because it takes a
+        lot of time to setup the initial instance. The bad part about this is
+        it might be a bit time consuming to debug and reproduce failures.
+        """
```
have you considered using 'moto' or 'vcr' like in the api_v2 tests instead of going all the way to spin up/down a real instance?
I will try not to oversell vcr, but I find it really handy as it produces a local stash of yaml requests/responses which are they consumed  by subsequent runs of the same test suite (as long as those files are available locally)
I find it speedups refactoring of things a lot when you've ran that particular test(s) at least once.
And love the hability to disable all the mocking of the AWS/Rackspace endpoints by just removing the cached yaml request/responses and let the tests run a full acceptance test suite.

go on, buy it,
you won't find fish fresher than mine

sarum90 replied:
VCR looked really awesome, I really should add that back in.... my one concern  was faking out the ssh check... I do really want to actually check SSH, I wonder if there is a way to ask VCR, Are you recording or playing back?

Azulinho replied:
never used this, but have it on my starred list of things that I thought could be useful one day (hoarder)
https://github.com/ncouture/MockSSH/blob/master/README.md

Azulinho replied:
if you do, I will probably steal the examples from you

sarum90 replied:
Yea, I could just mock ssh, but I was thinking that I actually do want to ensure we can ssh to instances that are spun up by this code. So... I want VCR to actually record the results of the first "ssh -i keyname user@ip" command, and then play it back on subsequent runs of the test.

Heh, so I have a terrible idea... I might code it up and see if you think it's as terrible as I think it is...

VCR can patch a "custom HTTPConnection" class... So I can make an "ExecCommand" class that implements the same interface as "HTTPConnection" (even though it makes no sense). Specifically on "request" it will actually execute a command (using check_output as a first pass), and return the output from the command (Probably wrapped in an HTTPrResponse because... reasons.

Then, when we use VCR to patch it, the first time it will actually record the result, and subsequent times it would play it back.

... seems sketchy enough that I might do this in a follow up. Hope that's okay - I think I'd rather have merged slow tests with a follow up coming than sync a couple more hours into this PR. But it does sound fun to try.

Azulinho replied:
works for me

Azulinho commented on bookshelf/tests/api_v3/test_cloud.py:
```diff
@@ -0,0 +1,339 @@
+            ami='ami-d440a6e7',
+            key_filename=credentials["ssh_keys"]["aws"]["private_key_file"],
+            key_pair=credentials["aws"]["keyname"],
+            instance_type='t2.medium',
```
t2.micro please, they're cheaper and these tests don't really do anything that requires anything larger anyway
Or even a t2.nano, maybe this value could be imported from the yaml file. 

http://www.huggard.org.uk/wp-content/uploads/2015/06/Tesco-Logo-Source.jpg

sarum90 replied:
Will do.

Azulinho commented:
brilliant fish curry here mate,
Consider the points I raised and address them if you agree with them
or merge otherwise

sarum90 commented:
Thanks, Just reviewed comments for now, will address in the branch and merge once I get in the office.

sarum90 committed:
Updated test from review.

This included making it so the tests can optionally take configuration values.
This functionality is used to let whomever runs the test choose where and on
what sized machine they want their tests to run.

sarum90 commented:
Thanks for the review!

None of the tests are actually run on travis, and travis doesn't have any tests for the README so I think I'm just going to merge as is, thanks!
